### PR TITLE
Switch to binary QR code

### PR DESCRIPTION
**Pull request by maxammann**
* Switched from base64 qr codes to binary ones
* Heavily adjusted how qr codes are generated in the administration
* Fixes things here and there

### CODE REVIEW

**Comment by maxammann** [comment]
I will test this after this is merged, so I can perform the tests directly on produciton.

Basic qr scanning is tested on Android and iOS.

**Comment by maxammann** [comment]
Untested: Generating and scanning verification codes

File: administration/src/cards/CardBlueprint.ts [hunk 1]
```diff
@@ -51,16 +52,17 @@ export class CardBlueprint {
...
 
-  sizeOfProtobuf(): number {
-    return new QrCode({ qrCode: { value: this.generateActivationCode(), case: 'dynamicActivationCode' } }).toBinary()
-      .length
+    const dynamicCode = new QrCode({
+      qrCode: { value: this.generateActivationCode(), case: 'dynamicActivationCode' },
+    }).toBinary()
+
+    const staticCode = new QrCode({
+      qrCode: { value: this.generateStaticVerificationCode(), case: 'dynamicActivationCode' },
```
Comment (michael-markl) [inline administration/src/cards/CardBlueprint.ts:62]:
```suggestion
      qrCode: { value: this.generateStaticVerificationCode(), case: 'staticVerificationCode' },
```
Wild, that this does not produce a typescript error. Can you check why?

File: administration/src/util/uint8ArrayToPaddedBase64.ts [hunk 2]
```diff
@@ -1,9 +1,9 @@
-const uint8ArrayToBase64 = function (u8: Uint8Array) {
+const uint8ArrayToPaddedBase64 = function (u8: Uint8Array) {
```
Comment (michael-markl) [inline administration/src/util/uint8ArrayToPaddedBase64.ts:1]: This function already exists in the `base64.ts` file. I think we can omit `padded` here. It's the default everywhere.

File: administration/src/components/cards/CreateCardsController.tsx [hunk 3]
```diff
@@ -34,6 +29,17 @@ const CreateCardsController = () => {
...
+  const projectConfig = useContext(ProjectConfigContext)
+  const client = useApolloClient()
+  const appToaster = useAppToaster()
+
+  const [cardBlueprints, setCardBlueprints] = useState<CardBlueprint[]>([projectConfig.createEmptyCard(region)])
+  const [state, setState] = useState(CardActivationState.input)
+
   const confirm = async () => {
     try {
       setState(CardActivationState.loading)
```
Comment (michael-markl) [inline administration/src/components/cards/CreateCardsController.tsx:45]: Can you adjust the `ehrenamtskarten.pdf` file name below (to something like `berechtigungskarten.pdf`?)

File: administration/src/util/qrcode.ts [hunk 4]
```diff
@@ -1,37 +1,184 @@
...
+const DEFAULT_QUIET_ZONE_SIZE = 4 // pt
+
+// Level 8 with EC of M gives 152 bytes
+// Level 7 with EC of L gives 154 bytes
+// Level 8 seems appropriate right now.
+export const DEFAULT_VERSION: QRCodeVersion = QRCodeVersion.getVersionForNumber(8)
+// From EC L to M we have a double of EC capability: 7% -> 15%
+export const DEFAULT_ERROR_CORRECTION: QRCodeDecoderErrorCorrectionLevel = QRCodeDecoderErrorCorrectionLevel.M
+
+function calculateBitsNeeded(
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:26]: Can you add a comment explaining what kind of size is calculated here?

File: administration/src/util/qrcode.ts [hunk 5]
```diff
@@ -1,37 +1,184 @@
...
+}
+
+function willFit(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel
+): boolean {
+  // In the following comments, we use numbers of Version 7-H.
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:42]: I thought we use 8-M?

File: administration/src/util/qrcode.ts [hunk 6]
```diff
@@ -1,37 +1,184 @@
...
+  // numBytes = 196
+  const numBytes = version.getTotalCodewords()
+  // getNumECBytes = 130
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:45]:
```suggestion
  // numECBytes = 130
```

File: administration/src/util/qrcode.ts [hunk 7]
```diff
@@ -1,37 +1,184 @@
...
+  const ecBlocks = version.getECBlocksForLevel(ecLevel)
+  const numEcBytes = ecBlocks.getTotalECCodewords()
+  // getNumDataBytes = 196 - 130 = 66
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:48]:
```suggestion
  // numDataBytes = 196 - 130 = 66
```

File: administration/src/util/qrcode.ts [hunk 8]
```diff
@@ -1,37 +1,184 @@
...
+  const numDataBytes = numBytes - numEcBytes
+  const totalInputBytes = (calculateBitsNeeded(mode, headerBitsSize, dataBitsSize, version) + 7) / 8
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:50]: Should we `Math.ceil` this? Also what is this magic 7?
```suggestion
  const totalInputBytes = Math.ceil((calculateBitsNeeded(mode, headerBitsSize, dataBitsSize, version) + 7) / 8)
```

File: administration/src/util/qrcode.ts [hunk 9]
```diff
@@ -1,37 +1,184 @@
...
+function willFit(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel
+): boolean {
+  // In the following comments, we use numbers of Version 7-H.
+  // numBytes = 196
+  const numBytes = version.getTotalCodewords()
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:44]: Should this be called `maxNumBytes`?

File: administration/src/util/qrcode.ts [hunk 10]
```diff
@@ -1,37 +1,184 @@
...
+  // getNumECBytes = 130
+  const ecBlocks = version.getECBlocksForLevel(ecLevel)
+  const numEcBytes = ecBlocks.getTotalECCodewords()
+  // getNumDataBytes = 196 - 130 = 66
+  const numDataBytes = numBytes - numEcBytes
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:49]:
```suggestion
  const maxNumDataBytes = maxNumBytes - numEcBytes
```

File: administration/src/util/qrcode.ts [hunk 11]
```diff
@@ -1,37 +1,184 @@
...
+  return headerBitsSize + mode.getCharacterCountBits(version) + dataBitsSize
+}
+
+function willFit(
+  mode: QRCodeMode,
+  headerBitsSize: number,
+  dataBitsSize: number,
+  version: QRCodeVersion,
+  ecLevel: QRCodeDecoderErrorCorrectionLevel
+): boolean {
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:41]: Can you add a link that explains these calculations?

File: administration/src/util/qrcode.ts [hunk 12]
```diff
@@ -1,37 +1,184 @@
...
+  // getNumECBytes = 130
+  const ecBlocks = version.getECBlocksForLevel(ecLevel)
+  const numEcBytes = ecBlocks.getTotalECCodewords()
+  // getNumDataBytes = 196 - 130 = 66
+  const numDataBytes = numBytes - numEcBytes
+  const totalInputBytes = (calculateBitsNeeded(mode, headerBitsSize, dataBitsSize, version) + 7) / 8
+  return numDataBytes >= totalInputBytes
+}
+
+function chooseMaskPattern(
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:54]: This is definitely quite low level :D Maybe you can provide a link with documentation/spec

File: administration/src/util/qrcode.ts [hunk 13]
```diff
@@ -1,37 +1,184 @@
...
+    MaskUtil.applyMaskPenaltyRule1(matrix) +
+    MaskUtil.applyMaskPenaltyRule2(matrix) +
+    MaskUtil.applyMaskPenaltyRule3(matrix) +
+    MaskUtil.applyMaskPenaltyRule4(matrix)
+  )
+}
+
+function createHeader(mode: QRCodeMode) {
+  // This will store the header information, like mode and
+  // length, as well as "header" segments like an ECI segment.
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:87]: What's an ECI segment? Please add to the comment

File: administration/src/util/qrcode.ts [hunk 14]
```diff
@@ -1,37 +1,184 @@
...
+  const headerBits = new BitArray()
+
+  // Do not append ECI segment, because we do not define the encoding of the QR code
+
+  // (With ECI in place,) Write the mode marker
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:92]: Not sure what these comments should tell me.

File: administration/src/util/qrcode.ts [hunk 15]
```diff
@@ -1,37 +1,184 @@
...
+export function encodeQRCode(content: Uint8Array): QRCode {
+  // Pick an encoding mode appropriate for the content.
+  const mode: QRCodeMode = QRCodeMode.BYTE
+
+  const headerBits = createHeader(mode)
+
+  // Collect data within the main segment, separately, to count its size if needed. Don't add it to
+  // main payload yet.
+  const dataBits = new BitArray()
+  for (let i = 0, length = content.length; i !== length; i++) {
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:119]:
```suggestion
  for (let i = 0; i < content.length; i++) {
```

File: administration/src/util/qrcode.ts [hunk 16]
```diff
@@ -1,37 +1,184 @@
...
+    const b = content[i]
+    dataBits.appendBits(b, 8)
+  }
+
+  let version: QRCodeVersion = DEFAULT_VERSION
+  let ecLevel: QRCodeDecoderErrorCorrectionLevel = DEFAULT_ERROR_CORRECTION
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:125]:
```suggestion
  const version: QRCodeVersion = DEFAULT_VERSION
  const ecLevel: QRCodeDecoderErrorCorrectionLevel = DEFAULT_ERROR_CORRECTION
```
(If possible)

File: frontend/lib/identification/activation_workflow/activation_code_parser.dart [hunk 17]
```diff
@@ -22,7 +24,7 @@ class QRCodeInvalidFormatException extends QrCodeParseException {
 class ActivationCodeParser {
   const ActivationCodeParser();
 
-  DynamicActivationCode parseQrCodeContent(String rawBase64Content) {
+  DynamicActivationCode parseQrCodeContent(Uint8List rawBase64Content) {
     final QrCode qrCode = rawBase64Content.parseQRCodeContent();
```
Comment (michael-markl) [inline frontend/lib/identification/activation_workflow/activation_code_parser.dart:28]: I guess the variable should not be called `rawBase64Content` anymore?

File: frontend/lib/identification/card_detail_view/verification_code_view.dart [hunk 18]
```diff
@@ -61,9 +61,12 @@ class _VerificationCodeViewState extends State<VerificationCodeView> {
                   children: [
                     Padding(
                       padding: EdgeInsets.all(padding),
-                      child: QrImage(
-                        data: const QrCodeUtils().createDynamicVerificationQrCodeData(activationCode, otpCode.code),
-                        version: QrVersions.auto,
+                      child: qr.QrImage.withQr(
+                        qr: qr.QrCode.fromUint8List(
+                          data: const QrCodeUtils().createDynamicVerificationQrCodeData(activationCode, otpCode.code),
+                          errorCorrectLevel: qr.QrErrorCorrectLevel.L,
```
Comment (michael-markl) [inline frontend/lib/identification/card_detail_view/verification_code_view.dart:67]: Don't we want to use Error Correction M here as well? Why not use auto ?

File: frontend/lib/identification/verification_workflow/verification_qr_content_parser.dart [hunk 19]
```diff
@@ -12,16 +12,14 @@ class VerificationParseException extends QrCodeParseException {
       : super(internalMessage);
 }
 
-extension QRParsing on String {
+extension QRParsing on Uint8List {
```
Comment (michael-markl) [inline frontend/lib/identification/verification_workflow/verification_qr_content_parser.dart:15]: It feels wrong that this is an extension on Uint8List. Also, it seems this is not just a verification_qr_content_parser, but also a general qr_code_parser. Maybe you can move this functionality to qr_code_utils.

File: frontend/lib/identification/verification_workflow/verification_qr_scanner_page.dart [hunk 20]
```diff
@@ -60,7 +60,7 @@ class VerificationQrScannerPage extends StatelessWidget {
                   otp: otp,
                 ),
               );
-              final verificationCodeBase64 = const Base64Encoder().convert(verificationQrCode.writeToBuffer());
+              final verificationCodeBase64 = verificationQrCode.writeToBuffer();
```
Comment (michael-markl) [inline frontend/lib/identification/verification_workflow/verification_qr_scanner_page.dart:63]: I think we should rename this variable?

**Review by michael-markl — APPROVED** [review]

File: administration/src/util/uint8ArrayToPaddedBase64.ts (hunk 2 above)
Comment (michael-markl) [reply administration/src/util/uint8ArrayToPaddedBase64.ts:1]: Ah sorry, it isn't there (yet). I changed that in my pr #787 ... Still, I think we can drop `padded` :D

File: administration/src/util/qrcode.ts [hunk 21]
```diff
@@ -1,37 +1,184 @@
...
+
+  // Collect data within the main segment, separately, to count its size if needed. Don't add it to
+  // main payload yet.
+  const dataBits = new BitArray()
+  for (let i = 0, length = content.length; i !== length; i++) {
+    const b = content[i]
+    dataBits.appendBits(b, 8)
+  }
+
+  let version: QRCodeVersion = DEFAULT_VERSION
```
Comment (michael-markl) [inline administration/src/util/qrcode.ts:124]: Optimally, we would dynamically choose the lowest possible version, to make the code only as dense as necessary. (And maybe throw if it is higher than a maximum version).

Comment (maxammann) [reply administration/src/util/qrcode.ts:124]: I think choosing a reasonable fixed version is just fine for our usecase. We can estimate the size of the QR code pretty well. 150bytes fits pretty good right now.

Else we also have to think about reasonable upper bound.

File: frontend/lib/identification/verification_workflow/verification_qr_content_parser.dart (hunk 19 above)
Comment (maxammann) [reply frontend/lib/identification/verification_workflow/verification_qr_content_parser.dart:15]: We can do that in another PR. and generally refactor things.

File: frontend/lib/identification/card_detail_view/verification_code_view.dart (hunk 18 above)
Comment (maxammann) [reply frontend/lib/identification/card_detail_view/verification_code_view.dart:67]: Seems like there is no auto with binary QR codes: https://pub.dev/documentation/qr/2.1.0/qr/QrErrorCorrectLevel-class.html

File: frontend/lib/identification/activation_workflow/activation_code_parser.dart (hunk 17 above)
Comment (maxammann) [reply frontend/lib/identification/activation_workflow/activation_code_parser.dart:28]: Thanks, I'll rename!

File: frontend/lib/identification/verification_workflow/verification_qr_scanner_page.dart (hunk 20 above)
Comment (maxammann) [reply frontend/lib/identification/verification_workflow/verification_qr_scanner_page.dart:63]: Thanks, I'll rename!

File: administration/src/util/qrcode.ts (hunk 16 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:125]: We need a better linting....

File: administration/src/util/qrcode.ts (hunk 15 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:119]: This code is copied from Encoder.ts. I'll try to clarify this better by adding a comment: https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts

I do not want to change their code to keep compatability and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 14 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:92]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 13 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:87]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 12 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:54]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 11 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:41]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 10 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:49]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 9 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:44]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 8 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:50]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 7 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:48]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 6 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:45]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 5 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:42]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/util/qrcode.ts (hunk 4 above)
Comment (maxammann) [reply administration/src/util/qrcode.ts:26]: This code is copied from [Encoder.ts](https://github.com/zxing-js/library/blob/master/src/core/qrcode/encoder/Encoder.ts). I'll try to clarify this better by adding a comment.

I do not want to change their code to keep compatibility and avoid mistakes.

File: administration/src/components/cards/CreateCardsController.tsx (hunk 3 above)
Comment (maxammann) [reply administration/src/components/cards/CreateCardsController.tsx:45]: Will do!

File: administration/src/util/qrcode.ts (hunk 21 above)
Comment (michael-markl) [reply administration/src/util/qrcode.ts:124]: We could have version 8 as an upper bound. But I agree, that this is an optimization which is likely unnecessary.

**Pull request merged**